sky/shell/ui/animator.cc

```diff
@@ -7,151 +7,103 @@
 #include "base/bind.h"
 #include "base/message_loop/message_loop.h"
 #include "base/trace_event/trace_event.h"
-#include "flutter/services/rasterizer/rasterizer.mojom.h"
+#include "flutter/common/threads.h"
 
 namespace sky {
 namespace shell {
 
-const int kPipelineDepth = 3;
-
-Animator::Animator(const Engine::Config& config,
-                   rasterizer::RasterizerPtr rasterizer,
-                   Engine* engine)
-    : config_(config),
-      rasterizer_(rasterizer.Pass()),
+Animator::Animator(Rasterizer* rasterizer, Engine* engine)
+    : rasterizer_(rasterizer),
       engine_(engine),
-      outstanding_requests_(0),
-      did_defer_frame_request_(false),
-      engine_requested_frame_(false),
+      layer_tree_pipeline_(3),
+      pending_frame_sem_(1),
       paused_(false),
-      is_ready_to_draw_(false),
       weak_factory_(this) {}
 
-Animator::~Animator() {}
-
-void Animator::RequestFrame() {
-  if (engine_requested_frame_)
-    return;
-  TRACE_EVENT_ASYNC_BEGIN0("flutter", "Frame request pending", this);
-  engine_requested_frame_ = true;
-
-  DCHECK(!did_defer_frame_request_);
-  outstanding_requests_++;
-  TRACE_COUNTER1("flutter", "outstanding_requests_", outstanding_requests_);
-  if (outstanding_requests_ >= kPipelineDepth) {
-    TRACE_EVENT_INSTANT1("flutter", "Frame request deferred",
-                         TRACE_EVENT_SCOPE_THREAD, "outstanding_requests_",
-                         outstanding_requests_);
-    did_defer_frame_request_ = true;
-    return;
-  }
-
-  if (!AwaitVSync()) {
-    base::MessageLoop::current()->PostDelayedTask(
-        FROM_HERE,
-        base::Bind(&Animator::BeginFrame, weak_factory_.GetWeakPtr(), 0),
-        base::TimeDelta::FromMilliseconds(16));
-  }
-}
+Animator::~Animator() = default;
 
 void Animator::Stop() {
   paused_ = true;
 }
 
 void Animator::Start() {
-  Reset();
-  RequestFrame();
-}
+  if (!paused_) {
+    return;
+  }
 
-void Animator::Animate(mojo::gfx::composition::FrameInfoPtr frame_info) {
-  BeginFrame(frame_info->frame_time);
+  paused_ = false;
+  RequestFrame();
 }
 
 void Animator::BeginFrame(int64_t time_stamp) {
-  TRACE_EVENT_ASYNC_END0("flutter", "Frame request pending", this);
-  DCHECK(engine_requested_frame_);
-  DCHECK(outstanding_requests_ > 0);
-  DCHECK(outstanding_requests_ <= kPipelineDepth) << outstanding_requests_;
+  if (paused_) {
+    return;
+  }
 
-  engine_requested_frame_ = false;
+  pending_frame_sem_.Signal();
 
-  if (paused_) {
-    OnFrameComplete();
+  LayerTreePipeline::Producer producer = [&]() {
```

Comment:
`[&]` is banned by the style guide.  Please list the variables you're capturing explicitly.